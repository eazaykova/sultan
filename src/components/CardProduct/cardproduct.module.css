.container {
	width: 326px;
	height: auto;
	padding-left: 23px;
	padding-right: 21px;

	min-height: 100%;
	display: flex;
	flex-direction: column;

	border-radius: 10px;
	background: #FFFFFF;
	box-shadow: 0px 14px 39px -11px rgba(0, 0, 0, 0.06);
}

.box {
	max-width: 100%;
	height: 240px;
	display: flex;
	justify-content: center;
	align-items: flex-end;
}

.box img {
	max-width: 100%;
	max-height: 100%;
	margin-bottom: 5px;
}

.boxSize {
	display: flex;
	margin-bottom: 10px;
}

.body {
	flex: 1 1 auto;
	cursor: pointer;
}

.basic {
	display: flex;
	align-items: flex-end;
	margin: 15px 0 29px;
}

.title {
	margin-bottom: 20px;
}

.price {
	margin-right: 40px;
	align-self: center;
}

.modal {
	display: flex;
	flex-direction: column;
	align-items: center;
}

.close {
	display: flex;
	align-self: flex-end;
	margin: 30px 30px 0px 0px;
	cursor: pointer;
}

.modalTitle {
	font-weight: 500;
	font-size: 40px;
	line-height: 120%;
	color: #111111;

	display: flex;
	height: 300px;
	text-align: center;
	align-items: center;
}

@media (max-width: 320px) { 
	.modalTitle {
		font-size: 20px;
		padding: 0px 20px;
		height: 200px;
	}
	.container {
		width: 310px;
	}
}


