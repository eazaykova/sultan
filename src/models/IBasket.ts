export interface IBasket {
	id: number;
	value: number;
	cost: number;
	price: number;
}