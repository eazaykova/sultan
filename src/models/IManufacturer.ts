export interface IManufacturer {
	manufact: string | null;
	count: number;
}
