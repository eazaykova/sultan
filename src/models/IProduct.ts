export interface IProduct {
	id: number;
	image: string;
	title: string;
	type: string;
	size: string;
	barcode: string;
	manufacturer: string;
	brand: string;
	description: string;
	price: number;
	typeproduct: string[];
}