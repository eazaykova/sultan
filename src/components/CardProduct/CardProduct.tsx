import { IProduct } from "../../models";
import Button from "../ui/Button/Button";
import StringForm from "../ui/StringForm/StringForm";
import styles from "./cardproduct.module.css";
import { useNavigate } from "react-router-dom";
import basketIcon from '../../images/icon/basket.svg';
import ProductType from "../ui/ProductType/ProductType";
import ProductTitle from "../ui/ProductTitle/ProductTitle";
import ProductPrice from "../ui/ProductPrice/ProductPrice";
import Modal from "../../components/ui/Modal/Modal";
import closeIcon from "../../images/icon/close.svg";

import { basketSlice } from '../../store/reducers/BasketSlice';
import { useAppDispatch, useAppSelector } from '../../hooks/redux';
import { useState } from "react";

interface ProductProps {
	product: IProduct;
}

export function CardProduct({ product }: ProductProps) {
	let typeProd = '';
	for (let type of product.typeproduct) {
		typeProd = typeProd + ' / ' + type;
	}

	const navigate = useNavigate();

	function openCardAndPassValue() {
		localStorage.setItem("Cost", JSON.stringify(product.price));
		let value = product.barcode + '-' + product.id;
		navigate(`/productcard/${value}`);
	}

	const { addToBasket } = basketSlice.actions;
	const dispatch = useAppDispatch();
	const { basket } = useAppSelector(state => state.basketReducer);

	const [modalActive, setModalActive] = useState(false);
	const [message, setMessage] = useState('');

	function addToBasketWithAMessage() {
		if (basket.find(item => item.id === product.id)) {
			setModalActive(true);
			setMessage('Этот товар уже в корзине!');
		} else {
			dispatch(addToBasket({ id: product.id, value: 1, cost: product.price }));
			setMessage('Товар добавлен в корзину!');
			setModalActive(true);
		}
	}

	return (
		<div className={styles.container} key={product.id}>
			<div onClick={() => openCardAndPassValue()} className={styles.body}>
				<div className={styles.box}>
					<img src={product.image} alt={product.title} />
				</div>

				<div className={styles.boxSize} >
					<ProductType type={product.type} size={product.size} />
				</div>
				<div className={styles.title}>
					<ProductTitle classN='title' brand={product.brand} title={product.title} />
				</div>

				<div className={styles.string} >
					<StringForm str='Штрихкод:' value={product.barcode} />
					<StringForm str='Производитель:' value={product.manufacturer} />
					<StringForm str='Бренд:' value={product.brand} />
					<StringForm str='Тип:' value={typeProd.substring(2)} />
				</div>
			</div>

			<div className={styles.basic} >
				<div className={styles.price}>
					<ProductPrice classN='priceS' price={product.price} />
				</div>
				<div><Button onClick={() => addToBasketWithAMessage()} classN='first' name='В КОРЗИНУ' icon={basketIcon} alt='Basket' /></div>
			</div>

			<Modal active={modalActive} setActive={setModalActive} >
				<div className={styles.modal}>
					<div className={styles.close} onClick={() => setModalActive(false)}><img src={closeIcon} alt="Close" /></div>
					<div className={styles.modalTitle}>{message}</div>
				</div>
			</Modal>


		</div>
	)
}
