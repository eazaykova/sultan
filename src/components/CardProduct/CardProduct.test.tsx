import { CardProduct } from './CardProduct';
import { render } from '@testing-library/react';
import * as reduxHooks from '../../hooks/redux';
import { IProduct } from "../../models/IProduct";
import { MemoryRouter } from 'react-router-dom';


const product: IProduct = {
	id: 1,
	image: 'https://emksp.ru/files/d48/d48a4d60b53c2d6b818caaa764927893-fit-400x400.jpg',
	title: 'Косметическое мыло',
	type: 'г',
	size: '540',
	barcode: '4604049097014',
	manufacturer: 'Colgate-Palmolive',
	brand: 'PALMOLIVE',
	description: 'Представляем Вашему вниманию восхитительный набор Косметического мыла от Palmolive. Разнообразие ароматов никого не оставит равнодушным. Освежающее с летним арбузом содержит увлажняющие компоненты и придает мягкость вашей коже.',
	price: 488,
	typeproduct: ['Уход за руками', 'Уход за телом']
}

jest.mock('../../hooks/redux');
const mockedSelector = jest.spyOn(reduxHooks, 'useAppSelector');

describe('CardProduct', () => {
	test('Creating a product card', () => {
		mockedSelector.mockReturnValue(product);
		const component = render(
			<MemoryRouter>
				<CardProduct product={product} />
			</MemoryRouter>
		);
		expect(component).toMatchSnapshot();
	});
})