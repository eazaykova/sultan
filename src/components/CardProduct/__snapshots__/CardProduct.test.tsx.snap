// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardProduct Creating a product card 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="body"
        >
          <div
            class="box"
          >
            <img
              alt="Косметическое мыло"
              src="https://emksp.ru/files/d48/d48a4d60b53c2d6b818caaa764927893-fit-400x400.jpg"
            />
          </div>
          <div
            class="boxSize"
          >
            <div>
              <img
                alt="icon"
                class="imgG"
                src="g.svg"
              />
              <span
                class="span"
              >
                540
              </span>
              <span
                class="span"
              >
                г
              </span>
            </div>
          </div>
          <div
            class="title"
          >
            <p
              class="title"
            >
              <span>
                PALMOLIVE
              </span>
               
              Косметическое мыло
            </p>
          </div>
          <div
            class="string"
          >
            <p
              class="string"
            >
              Штрихкод:
              <span
                class="value"
              >
                4604049097014
              </span>
            </p>
            <p
              class="string"
            >
              Производитель:
              <span
                class="value"
              >
                Colgate-Palmolive
              </span>
            </p>
            <p
              class="string"
            >
              Бренд:
              <span
                class="value"
              >
                PALMOLIVE
              </span>
            </p>
            <p
              class="string"
            >
              Тип:
              <span
                class="value"
              >
                 Уход за руками / Уход за телом
              </span>
            </p>
          </div>
        </div>
        <div
          class="basic"
        >
          <div
            class="price"
          >
            <span
              class="priceS"
            >
              488
               ₸
            </span>
          </div>
          <div>
            <button
              class="first"
              data-testid="btn-basket"
              type="button"
            >
               
              В КОРЗИНУ
               
              <img
                alt="Basket"
                src="basket.svg"
              />
               
            </button>
          </div>
        </div>
        <div
          class="modal"
        >
          <div
            class="content"
          >
            <div
              class="modal"
            >
              <div
                class="close"
              >
                <img
                  alt="Close"
                  src="close.svg"
                />
              </div>
              <div
                class="modalTitle"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="body"
      >
        <div
          class="box"
        >
          <img
            alt="Косметическое мыло"
            src="https://emksp.ru/files/d48/d48a4d60b53c2d6b818caaa764927893-fit-400x400.jpg"
          />
        </div>
        <div
          class="boxSize"
        >
          <div>
            <img
              alt="icon"
              class="imgG"
              src="g.svg"
            />
            <span
              class="span"
            >
              540
            </span>
            <span
              class="span"
            >
              г
            </span>
          </div>
        </div>
        <div
          class="title"
        >
          <p
            class="title"
          >
            <span>
              PALMOLIVE
            </span>
             
            Косметическое мыло
          </p>
        </div>
        <div
          class="string"
        >
          <p
            class="string"
          >
            Штрихкод:
            <span
              class="value"
            >
              4604049097014
            </span>
          </p>
          <p
            class="string"
          >
            Производитель:
            <span
              class="value"
            >
              Colgate-Palmolive
            </span>
          </p>
          <p
            class="string"
          >
            Бренд:
            <span
              class="value"
            >
              PALMOLIVE
            </span>
          </p>
          <p
            class="string"
          >
            Тип:
            <span
              class="value"
            >
               Уход за руками / Уход за телом
            </span>
          </p>
        </div>
      </div>
      <div
        class="basic"
      >
        <div
          class="price"
        >
          <span
            class="priceS"
          >
            488
             ₸
          </span>
        </div>
        <div>
          <button
            class="first"
            data-testid="btn-basket"
            type="button"
          >
             
            В КОРЗИНУ
             
            <img
              alt="Basket"
              src="basket.svg"
            />
             
          </button>
        </div>
      </div>
      <div
        class="modal"
      >
        <div
          class="content"
        >
          <div
            class="modal"
          >
            <div
              class="close"
            >
              <img
                alt="Close"
                src="close.svg"
              />
            </div>
            <div
              class="modalTitle"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
